src/braket/default_simulator/openqasm/program_context.py

```diff
@@ -113,29 +115,73 @@ def _(self, identifier: IndexedIdentifier) -> tuple[int]:
         corresponding to the elements referenced by the indexed identifier.
         """
         name = identifier.name.name
+        indices = self.get_qubit_indices(identifier)
+        primary_index = indices[0]
+
+        if isinstance(primary_index, (IntegerLiteral, SymbolLiteral)):
+            target = (self[name][0] + primary_index.value,)
+        elif isinstance(primary_index, RangeDefinition):
+            target = tuple(np.array(self[name])[convert_range_def_to_slice(primary_index)])
+        # Discrete set
+        else:
+            index_list = convert_discrete_set_to_list(primary_index)
+            target = tuple([self[name][0] + index for index in index_list])
+
+        if len(indices) == 2:
+            # used for gate calls on registers, index will be IntegerLiteral
+            secondary_index = indices[1].value
+            target = (target[secondary_index],)
+
+        # validate indices manually, since we use addition instead of indexing to
+        # accommodate symbolic indices
+        for q in target:
+            if isinstance(q, int) and (relative_index := q - self[name][0]) >= len(self[name]):
+                raise IndexError(
+                    f"Index {relative_index} out of bounds for qubit '{name}' "
+                    f"with size {len(self[name])}"
+                )
+        return target
+
+    @staticmethod
+    def get_qubit_indices(
+        identifier: IndexedIdentifier,
+    ) -> list[IntegerLiteral | RangeDefinition | DiscreteSet]:
         primary_index = identifier.indices[0]
         if isinstance(primary_index, list):
             if len(primary_index) != 1:
                 raise IndexError("Cannot index multiple dimensions for qubits.")
             primary_index = primary_index[0]
-        if isinstance(primary_index, IntegerLiteral):
-            target = (self[name][primary_index.value],)
-        elif isinstance(primary_index, RangeDefinition):
-            target = tuple(np.array(self[name])[convert_range_def_to_slice(primary_index)])
-        # Discrete set
-        else:
-            target = tuple(np.array(self[name])[convert_discrete_set_to_list(primary_index)])
 
         if len(identifier.indices) == 1:
-            return target
+            return [primary_index]
         elif len(identifier.indices) == 2:
             # used for gate calls on registers, index will be IntegerLiteral
-            secondary_index = identifier.indices[1][0].value
-            return (target[secondary_index],)
+            secondary_index = identifier.indices[1][0]
+            return [primary_index, secondary_index]
         else:
             raise IndexError("Cannot index multiple dimensions for qubits.")
 
+    def get_indices_length(
+        self,
+        indices: Sequence[IntegerLiteral | SymbolLiteral | RangeDefinition | DiscreteSet],
+    ):
```

Comment:
This method allows us to get the length of a qubit register that may have unbound symbols. This will be more crucial (and better tested) when the range transformation can actually be used with a symbolic "loop" construct in the IR. For now, if this returns something symbolic, the interpreter can't use it.